### PR TITLE
feat: Support FLOAT32 datatype

### PR DESCRIPTION
**Pull request by SandeepTuniki**
This PR adds support for FLOAT32 datatype in Ruby Spanner.

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/convert.rb [hunk 1]
```diff
@@ -211,6 +211,20 @@ def grpc_value_to_object value, type
               else
                 value.number_value
               end
+            when :FLOAT32
+              if value.kind == :string_value
```
Comment (NivedhaSenthil) [inline google-cloud-spanner/lib/google/cloud/spanner/convert.rb:215]: seems this block is repeated for float 32 and 64 can we pull it out ?

Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/convert.rb:201]: You're right. I missed this. Fixed it now.

File: google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb [hunk 2]
```diff
@@ -0,0 +1,228 @@
+# Copyright 2024 Google LLC
+#
```
Comment (NivedhaSenthil) [inline google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb:2]: We may have to add params test for float32 as well and ensure it is handled in code.

Comment (NivedhaSenthil) [reply google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb:2]: And if user does not specify the param type what would take precedence Float32 or 64 ?

**Review by NivedhaSenthil — CHANGES_REQUESTED** [review]

File: google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb (hunk 2 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb:2]: > We may have to add params test for float32 as well and ensure it is handled in code.

Missed this, thanks for spotting it. I added the tests for params in the `params/float32_test.rb` file.

Comment (SandeepTuniki) [reply google-cloud-spanner/acceptance/spanner/client/types/float32_test.rb:2]: > And if user does not specify the param type what would take precedence Float32 or 64 ?

It defaults to float64 (there is already an existing test for that).

**Review by NivedhaSenthil — APPROVED** [review]

**Comment by SandeepTuniki** [comment]
The failing acceptance is a flaky Backup operation, and is unrelated to this PR.

File: google-cloud-spanner/acceptance/spanner/client/params/float32_test.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "spanner_helper"
+
+describe "Spanner Client", :params, :float32, :spanner do
+  let(:db) { spanner_client }
+
+  it "queries and returns a float32 parameter" do
+    results = db.execute_query "SELECT @value AS value", params: { value: 12.0 }, types: { value: :FLOAT32 }
```
Comment (arawind) [inline google-cloud-spanner/acceptance/spanner/client/params/float32_test.rb:21]: As discussed, please add a new test for `SELECT CAST(@value AS FLOAT32) AS value` with the type code not set. This should return a `FLOAT32: 12.0` (approximately)

**Review by arawind — APPROVED** [review]
LGTM, thanks!

File: google-cloud-spanner/acceptance/spanner/client/params/float32_test.rb (hunk 3 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/acceptance/spanner/client/params/float32_test.rb:21]: Thanks for spotting this case. Added another test (just below this test).

**Pull request merged**